File: 백준/Gold/14502. 연구소/연구소.cc
#include <bits/stdc++.h>
using namespace std;

int a[10][10], visited[10][10], n, m, ret;
vector<pair<int, int>> wallList, virusList;
const int dy[] = {-1, 0, 1, 0};
const int dx[] = {0, 1, 0, -1};

bool InRange(int y, int x) {
    return y >= 0 && y < n && x >= 0 && x < m;
}

void dfs(int y, int x) {
    visited[y][x] = 1;
    for(int i = 0; i < 4; i++) {
        int ny = y + dy[i];
        int nx = x + dx[i];
        
        if(InRange(ny, nx) && a[ny][nx] == 0 && visited[ny][nx] == 0) {
            dfs(ny, nx);
        }
    }
    return;
}

int solve() {
    int cnt = 0;
    fill(&visited[0][0], &visited[0][0] + 10 * 10, 0);
    for(pair<int, int> b : virusList) {
        dfs(b.first, b.second);
    }
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            if(a[i][j] == 0 && visited[i][j] == 0) cnt++;
        }
    }
    return cnt;
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    cin >> n >> m;
    
    for(int i = 0; i < n; i++) {
        for(int j = 0; j < m; j++) {
            cin >> a[i][j];
            if(a[i][j] == 0) wallList.push_back({i, j});
            else if(a[i][j] == 2) virusList.push_back({i, j});
        }
    }
    
    for(int i = 0; i < wallList.size(); i++) {
        for(int j = i + 1; j < wallList.size(); j++) {
            for(int k = j + 1; k < wallList.size(); k++) {
                a[wallList[i].first][wallList[i].second] = 1;
                a[wallList[j].first][wallList[j].second] = 1;
                a[wallList[k].first][wallList[k].second] = 1;
                ret = max(ret, solve());
                a[wallList[i].first][wallList[i].second] = 0;
                a[wallList[j].first][wallList[j].second] = 0;
                a[wallList[k].first][wallList[k].second] = 0;
            }
        }
    }
    
    cout << ret << '\n';
    return 0;
}

File: 백준/Gold/14502. 연구소/연구소.swift
import Foundation

let input = readLine()!.components(separatedBy: " ").map { Int($0)! }
let n = input[0], m = input[1]

var arr = [[Int]]()
var virusList = [(Int, Int)]()
var wallList = [(Int, Int)]()
var ret = 0
var visited: [[Int]] = []
let dy = [-1, 0, 1, 0]
let dx = [0, 1, 0, -1]

for _ in 0..<n {
    let line = readLine()!.components(separatedBy: " ").map { Int($0)! }
    arr.append(line)
}

for i in 0..<n {
    for j in 0..<m {
        if arr[i][j] == 2 { virusList.append((i, j)) }
        else if arr[i][j] == 0 { wallList.append((i, j)) }
    }
}

for i in 0..<wallList.count {
    for j in 0..<i {
        for k in 0..<j {
            arr[wallList[i].0][wallList[i].1] = 1
            arr[wallList[j].0][wallList[j].1] = 1
            arr[wallList[k].0][wallList[k].1] = 1
            ret = max(solve(), ret)
            arr[wallList[i].0][wallList[i].1] = 0
            arr[wallList[j].0][wallList[j].1] = 0
            arr[wallList[k].0][wallList[k].1] = 0
        }
    }
}

print(ret)

func solve() -> Int {
    visited = Array(repeating: Array(repeating: 0, count: m), count: n)
    
    for virus in virusList {
        visited[virus.0][virus.1] = 1
        go(virus.0, virus.1)
    }
    
    var cnt = 0
    for i in 0..<n {
        for j in 0..<m {
            if arr[i][j] == 0 && visited[i][j] == 0 {
                cnt += 1
            }
        }
    }
    return cnt
}

func go(_ y: Int, _ x: Int) {
    for i in 0..<4 {
        let ny = y + dy[i]
        let nx = x + dx[i]
        
        if ny < 0 || ny >= n || nx < 0 || nx >= m { continue }
        
        if (arr[ny][nx] == 0 || arr[ny][nx] == 2) && visited[ny][nx] == 0 {
            visited[ny][nx] = 1
            go(ny, nx)
        }
    }
}